gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java

```diff
@@ -310,4 +325,30 @@ protected boolean needsEviction(final String token) throws UnknownTokenException
     return tokenExpirations.keySet().stream().collect(Collectors.toList());
   }
 
+  /**
+   * A function that returns the JWT token expiration. This is only called when
+   * gateway.knox.token.permissive.failure.enabled property is set to true.
+   *
+   * @param token token to be verified and saved
+   */
+  protected long getJWTTokenExpiration(final String token) {
+    if (!isValidIdentifier(token)) {
+      throw new IllegalArgumentException("Token data cannot be null.");
+    }
+    JWT jwt;
+    try {
+      jwt = new JWTToken(token);
+    } catch (final ParseException e) {
+      log.errorParsingToken(e.toString());
+      throw new IllegalArgumentException(e);
```

Comment:
So this throws a new exception? Where does this exception end up going? I think this will convert an `UnknownTokenException` to a runtime `IllegalArgumentException` and so this will cause weird failures up the stack?

Reply:
True, that would be ugly, but we only throw two types of exceptions here `UnknownTokenException ` and `IllegalArgumentException`. `ParseException` is not a Runtime exception so we need to wrap it up in one of the two types, which leaves us with `IllegalArgumentException`. The stack trace will atleast let users know what the original error was.

Reply:
I **do not** see this converting an UnknownTokenException to an IllegalArgumentException. Am I missing something?

Reply:
So its subtle. `getJWTTokenExpiration` is called from the following code:

```
 try {
      validateToken(token);
    } catch (final UnknownTokenException e) {
      /* if token permissiveness is enabled we check JWT token expiration when the token state is unknown */
      if (permissiveFailureEnabled && StringUtils
          .containsIgnoreCase(e.toString(), "Unknown token")) {
        return getJWTTokenExpiration(token);
      } else {
        throw e;
      }
    }
```

What this means is that originally the `validateToken` call would throw `UnknownTokenException`. With this change, the `UnknownTokenException` is caught and `getJWTTokenExpiration` is called. If `getJWTTokenExpiration` fails, the code throws `IllegalArgumentException`

So it has the same effect of converting `UnknownTokenException` to `IllegalArgumentException` based on the settings introduced.

Reply:
I see what you're saying, but really the UnknownTokenException is being handled, and the fact that the token is actually invalid is a different error.

Reply:
Right but the method signature where all this is called is:

`public long getTokenExpiration(final String token) throws UnknownTokenException {`

So why would we throw an `IllegalArgumentException`? Its an unchecked runtimeexception so do we know how that bubbles back up to the caller? Is every user of `DefaultTokenStateService#getTokenExpiration` going to know to check for `IllegalArgumentException`?

Reply:
So to fix this I can think of either of two ways 

1. We would need to update the method signature for `getTokenExpiration()` to throw `UnknownTokenException` and `IllegalArgumentException`.  Now this change would cause updating a lot of method signatures which I am hesitant to do as it affects a lot of parts that we know work well. 
2. `getJWTTokenExpiration()` method can throw `UnknownTokenException`. We do not need to check if the token is not null here (since it was done by calling method currently using it) so we can get rid of one of two `IllegalArgumentException`. And then wrap the `ParseException` into `UnknownTokenException` which I am not sure is right.

@risdenk @pzampino thoughts ?

Reply:
What if, rather than throwing the IllegalArgumentException, a value (e.g., 0) was returned? This would be handled as treating the token as if it is expired, which is primarily used to determine if it is (still) valid; We might have to change the resulting log message(s), but I think treating it as invalid would work.

Reply:
There could be some case where 0 might be a legal value. On the same lines I can just make return type [OptionalInt](https://docs.oracle.com/javase/8/docs/api/java/util/OptionalInt.html).

Reply:
You could also just make it [Long](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html), and return null in the case where it cannot be parsed. The OptionalLong might be better though.

Reply:
Cool, will do.